### PR TITLE
Idle CPU usage improvements

### PR DESCRIPTION
**Pull request by fedux**
This PRs simplifies mutex and threads, fixes some unprotected members and includes a few changes to improve CPU usage when idle.

Part of a solution to https://github.com/nzbget/nzbget/issues/351

### CODE REVIEW

**Comment by hugbug** [comment]
That's a brilliant work. I have some questions which I'll post inline.

File: daemon/queue/DownloadInfo.h [hunk 1]
```diff
@@ -226,7 +226,7 @@ class FileInfo
 	int m_completedArticles = 0;
 	bool m_outputInitialized = false;
 	CString m_outputFilename;
-	std::unique_ptr<Mutex> m_outputFileMutex;
+	Mutex m_outputFileMutex;
```
Comment (hugbug) [inline daemon/queue/DownloadInfo.h]: That creates a mutex for each file (for example rar-file) in download queue. One nzb-item may contain hundreds of files. With a hundred of nzb-files in queue many thousands of file-objects may exist. Are you sure it is OK to have thousands of mutexex created?

The previous implementation of Mutex were using system objects and I didn't want to create too many of system objects. How lightweight std::mutex is? Is there any part of C++ specification saying the mutexes are lightweight enough to be created in thousands without worries?

File: daemon/util/Thread.cpp [hunk 2]
```diff
@@ -41,29 +42,22 @@ void Thread::Start()
...
-	pthread_attr_setdetachstate(&m_attr, PTHREAD_CREATE_DETACHED);
-	pthread_attr_setinheritsched(&m_attr, PTHREAD_INHERIT_SCHED);
-	m_running = !pthread_create(&m_threadObj, &m_attr, Thread::thread_handler, (void *) this);
-	pthread_attr_destroy(&m_attr);
-#endif
+	// wait for the ready signal
+	futureReady.wait();
+
+	// save the native handler to be able to cancel (Kill) the thread and then detach
+	m_threadObj = t.native_handle();
```
Comment (hugbug) [inline daemon/util/Thread.cpp:59]: At this moment the thread may be already terminated (if it was short-lived). If it was auto-destroying (`m_autoDestroy`) then `this` is already destroyed and we can't access `m_threadObj` anymore.

Would moving this line into thread's lambda function fix the issue?

----

> // save the native handler

A typo: handler -> handle.

Comment (hugbug) [reply daemon/util/Thread.cpp:59]: What is the advantage of using promise/future over a mutex?

Promise/future is a new thing to me and I'm trying to understand how they work and if they are better than manually managed mutexes. 

Promise/future has some kind of thread synchronisation under the hood as well, right? The old code was using one synchronisation (via mutex) to start the thread and increase `m_threadCount`. The new implementation has two synchronisations - one via promise/future and another one via atomic. Generally more synchronisations mean more CPU cache flushes and worse performance.

File: daemon/queue/DownloadInfo.h (hunk 1 above)
Comment (fedux) [reply daemon/queue/DownloadInfo.h]: Yeah, there is no problem with having many std::mutex instances. Just as a test I created a few millions without any problems. AFAIK the C++ specification doesn't deal with the implementation details.
I was also wondering why you have to create the FileInfo objects for Nzb files not active in the queue, any even maybe we can get rid of this mutex completely later.
In any case, if you still prefer to skip this commit then I can do that.

File: daemon/util/Thread.cpp (hunk 2 above)
Comment (fedux) [reply daemon/util/Thread.cpp:59]: > At this moment the thread may be already terminated (if it was short-lived). If it was auto-destroying (`m_autoDestroy`) then `this` is already destroyed and we can't access `m_threadObj` anymore.

OK, I see what you mean. I think for now I can keep the old mutex but use it only for ensuring this. I wonder if the need for the `m_autoDestroy` can be avoided but I still haven't looked at that part of the code yet.

> What is the advantage of using promise/future over a mutex?

It's higher level and let's the underlying implementation to use the best method to implement it. promise/future are useful to send results and states between threads and are meant to be used only once, which fits nice into this use case.

> The new implementation has two synchronisations - one via promise/future and another one via atomic. 

Using atomic is nice if the thread contention is low as it doesn't stop the threads (it locks but doesn't stop the waiting thread and switch context). In the previous implementation we had a few shared values not protected and using atomic is in my opinion a nice solution as it's more efficient than a mutex (mutexes use atomic operations internally) and thread contention should be very low for this cases.

File: daemon/queue/DownloadInfo.h (hunk 1 above)
Comment (hugbug) [reply daemon/queue/DownloadInfo.h]: > if you still prefer to skip this commit then I can do that.

Yes please. I'm not comfortable with thousands of (unused) mutexes hanging around.

Comment (fedux) [reply daemon/queue/DownloadInfo.h]: Changed.

File: daemon/util/Thread.h [hunk 3]
```diff
@@ -73,15 +74,15 @@ class Thread
...
 	HANDLE m_threadObj = 0;
 #else
 	pthread_t m_threadObj = 0;
 #endif
-	bool m_running = false;
-	bool m_stopped = false;
-	bool m_autoDestroy = false;
+	std::atomic_bool m_running{false};
+	std::atomic_bool m_stopped{false};
+	std::atomic_bool m_autoDestroy{false};
```
Comment (hugbug) [inline daemon/util/Thread.h]: `m_autoDestroy` is set before the thread is started and is never changed later; no need to protect.

As for `m_running` and `m_stopped` - is protecting here really necessary? The thread worker functions usually often check thread status via `IsStopped` to decide if the processing must be aborted. For example in article downloader it is checked after every (small) chunk of downloaded data - that maybe thousands times per second (in every download thread).

I thought the synchronisation can be omitted here since we work with a primitive type (bool) which is read/written with a single CPU operation (hopefully, yeah). Worst case scenario - the thread which reads the value may get the older value of the variable but that's not critical as it will eventually get the correct value somewhat later. Is this asking for troubles or a legitimate optimisation?

Comment (fedux) [reply daemon/util/Thread.h]: Ok, I will revert `m_autoDestroy` to bool. About the other two, in my opinion it's better to program safe than try to optimize something that doesn't really make a difference in performance, but it's your call if you want this to be reverted as well.

Comment (hugbug) [reply daemon/util/Thread.h]: > try to optimize something that doesn't really make a difference in performance

Performance was my concern exactly. In my short performance test on x86_64 the performance hasn't changed between before-PR and with-PR. I haven't tested on ARM as these tests are more difficult to perform there. I may do tests later though. I can't test other CPUs (MIPS, PowerPC) at all.

For now I suggest to use atomic (for `m_running` and ´m_stopped´).

File: daemon/util/Thread.cpp (hunk 2 above)
Comment (hugbug) [reply daemon/util/Thread.cpp:59]: > OK, I see what you mean. I think for now I can keep the old mutex but use it only for ensuring this.

Once this is fixed I can merge the PR.

I've tried this and it worked:

```C++
void Thread::Start()
{
	debug("Starting Thread");

	m_running = true;

	// NOTE: "m_threadMutex" ensures that "t" lives until the very end of the function
	Guard guard (m_threadMutex);

	// start the new thread
	std::thread t([&]{
		{
			// trying to lock "m_threadMutex", this will wait until function "Start()" is completed
			// and "t" is detached.
			Guard guard(m_threadMutex);
		}

		thread_handler();
	});

	// save the native handle to be able to cancel (Kill) the thread and then detach
	m_threadObj = t.native_handle();
	t.detach();
}
```

I've also tried to solve this with promise/future but I needed two pairs of promise/future. The solution with mutex is simpler.

File: daemon/feed/FeedCoordinator.h [hunk 4]
```diff
@@ -31,6 +31,7 @@
 #include "Observer.h"
 #include "Util.h"
 
+#include <condition_variable>
```
Comment (hugbug) [inline daemon/feed/FeedCoordinator.h]: For this and other places: please move all new includes into [nzbget.h](https://github.com/nzbget/nzbget/blob/develop/daemon/main/nzbget.h#L225).

This improves compiling times with precompiled headers.

Comment (fedux) [reply daemon/feed/FeedCoordinator.h]: done

File: daemon/util/Thread.cpp (hunk 2 above)
Comment (fedux) [reply daemon/util/Thread.cpp:59]: Updated

**Comment by hugbug** [comment]
Thanks for applying the changes and a special thank for updating the commits instead of adding new commits with small corrections. I like the clean history.

I have one more small thing - will post an inline comment.

File: daemon/postprocess/PrePostProcessor.cpp [hunk 5]
```diff
@@ -59,13 +59,25 @@ void PrePostProcessor::Run()
...
+		if (m_queuedJobs)
 		{
 			// check post-queue every 200 msec
 			CheckPostQueue();
+			usleep(200 * 1000);
+		}
+		else
+		{
+			// Wait until we get the stop signal or more jobs in the queue
+			std::unique_lock<std::mutex> lk(m_pauseMutex);
```
Comment (hugbug) [inline daemon/postprocess/PrePostProcessor.cpp]: Should probably be:

```C++
UniqueLock lk(m_pauseMutex);
```

Maybe also better to use name `guard` just like in all other places? Not sure about that though.

Consequently `m_pauseMutex` should be declared as `Mutex` instead of `std::mutex`.

And similarly in `FeedCoordinator.cpp/.h`.

File: daemon/postprocess/PrePostProcessor.h [hunk 6]
```diff
@@ -42,8 +42,10 @@ class PrePostProcessor : public Thread, public Observer
 	virtual void Update(Subject* caller, void* aspect) { DownloadQueueUpdate(aspect); }
 
 private:
-	int m_queuedJobs = 0;
+	std::atomic_uint m_queuedJobs{0};
```
Comment (hugbug) [inline daemon/postprocess/PrePostProcessor.h]: I use `int` whenever possible. Unsigned types considered dangerous.

Please replace with `std::atomic_int`. There are also few other places with `std::atomic_uint` in other units which should be replaced too. Thanks.

File: daemon/postprocess/PrePostProcessor.cpp (hunk 5 above)
Comment (fedux) [reply daemon/postprocess/PrePostProcessor.cpp]: > Maybe also better to use name guard just like in all other places? Not sure about that though.

Yeah, I thought about that but as the type is std::unique_lock lock I left it like that because it better matches the real type.

Changed the rest.

File: daemon/postprocess/PrePostProcessor.h (hunk 6 above)
Comment (fedux) [reply daemon/postprocess/PrePostProcessor.h]: changed

**Comment by hugbug** [comment]
I've run into an issue when compiling for linux installer package (cross compiling for many CPU architectures).

The compilation for **armel** breaks with:
```
daemon/main/nzbget.cpp:226:21: error: field ‘m_stopSignal’ has incomplete type ‘std::promise<void>’
  std::promise<void> m_stopSignal;
                     ^
In file included from daemon/main/nzbget.h:229:0,
                 from daemon/main/nzbget.cpp:22:
/mnt/build/toolchain/armel/output/host/usr/arm-buildroot-linux-uclibcgnueabi/include/c++/5.2.0/future:124:11: note: declaration of ‘class std::promise<void>’
     class promise;
           ^
```

After searching found [this explanation](https://lists.debian.org/debian-arm/2013/12/msg00000.html). In short (read the link for details):

> any C++11 program linked to libstdc++ which uses future facilities or transfer exceptions will fail to compile on armel.

Do you have an alternative solution without promise/future maybe? If you don't I would need to `ifdef` promise/future and need to extend the configure-script with detection of promise/future-support.

**Comment by hugbug** [comment]
I think we can use `std::condition_variable` here. I'll try this as an exercise.

**Comment by fedux** [comment]
Yes, it can be replaced by condition_variable, but it's a shame that it doesn't work for armel because that means we won't be able to use async or packed_task either probably. I'll see what I can find about it.

**Comment by fedux** [comment]
Apparently it's already fixed in gcc https://gcc.gnu.org/bugzilla/show_bug.cgi?id=64735

Which compiler version is failing?

**Comment by hugbug** [comment]
It's GCC 5.2.

Even though the problem seems to be fixed in GCC 7.0 I cannot easily drop support of older GCC versions. Armel is an important platform. It was a pain for some users when I migrated to C++11/14 with requirement for GCC 4.9. Even GCC 4.8 supported (but only when compiling in release mode) and I intentionally don't use some C++ features not available on GCC 4.8 in order to maintain compatibility.

Changing system requirements of NZBGet to higher GCC version is possible but it has to bring a big advantage to the project. A simple "there is one place where we could use a new compiler feature" isn't such case. The switch from C++98 to C++11/14 was such case.

If promise/future or async or packed_task will allow to for example increase performance to 10% then the switch may be worth it. If these features just allow to save a couple of lines of code then no.

Maintaining compatibility with many platforms is indeed painful and has costs.

**Comment by fedux** [comment]
debian has backported it to gcc-6 but I see your point. It's a shame because it's not really a requirement of a higher c++ standard but a bug in gcc.

In any case, it's not a big deal to use condition_variable instead.
